pkg/activator/activator.go

```diff
@@ -0,0 +1,272 @@
+	virtualServiceInformer cache.SharedIndexInformer,
+	autoscalerClient autoscaler.Client,
+	logger *zap.SugaredLogger,
+) Activator {
+	log := logger.With(zap.String("apiKind", userconfig.RealtimeAPIKind.String()))
+
+	act := &activator{
+		apiActivators:     make(map[string]*apiActivator),
+		readinessTrackers: make(map[string]*readinessTracker),
+		istioClient:       istioClient,
+		logger:            log,
+		autoscalerClient:  autoscalerClient,
+	}
+
+	virtualServiceInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    act.addAPI,
+			UpdateFunc: act.updateAPI,
+			DeleteFunc: act.removeAPI,
+		},
+	)
+
+	deploymentInformer.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: act.updateReadinessTracker,
+			UpdateFunc: func(_, newObj interface{}) {
+				act.updateReadinessTracker(newObj)
+			},
+			DeleteFunc: act.removeReadinessTracker,
+		},
+	)
+
+	return act
+}
+
+func (a *activator) Try(ctx context.Context, fn func() error) error {
+	apiNameValue := ctx.Value(APINameCtxKey)
+	apiName, ok := apiNameValue.(string)
+	if !ok || apiName == "" {
+		return errors.ErrorUnexpected("failed to get the api name from context")
+	}
+
+	act, err := a.getOrCreateAPIActivator(ctx, apiName)
+	if err != nil {
+		return err
+	}
+
+	tracker := a.getOrCreateReadinessTracker(apiName)
+
+	if act.inFlight() == 0 {
+		go a.awakenAPI(apiName)
+	}
+
+	return act.try(ctx, fn, tracker)
+}
+
+func (a *activator) getOrCreateAPIActivator(ctx context.Context, apiName string) (*apiActivator, error) {
+	a.activatorsMux.RLock()
+	act, ok := a.apiActivators[apiName]
+	a.activatorsMux.RUnlock()
+	if ok {
+		return act, nil
+	}
+
+	vs, err := a.istioClient.Get(ctx, workloads.K8sName(apiName), kmeta.GetOptions{})
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	maxQueueLength, maxConcurrency, err := concurrencyFromAnnotations(vs.Annotations)
+	if err != nil {
+		return nil, err
+	}
+
+	apiAct := newAPIActivator(maxQueueLength, maxConcurrency)
+
+	a.activatorsMux.Lock()
+	a.apiActivators[apiName] = apiAct
+	a.activatorsMux.Unlock()
+
+	return apiAct, nil
+}
+
+func (a *activator) getOrCreateReadinessTracker(apiName string) *readinessTracker {
+	a.trackersMux.RLock()
+	tracker, ok := a.readinessTrackers[apiName]
+	a.trackersMux.RUnlock()
+	if ok {
+		return tracker
+	}
+
+	tracker = newReadinessTracker()
+
+	a.trackersMux.Lock()
+	a.readinessTrackers[apiName] = tracker
+	a.trackersMux.Unlock()
+
+	return tracker
+}
+
+func (a *activator) addAPI(obj interface{}) {
+	apiMetadata, err := getAPIMeta(obj)
+	if err != nil {
+		a.logger.Errorw("error during virtual service informer add callback", zap.Error(err))
+		return
+	}
+
+	if apiMetadata.apiKind != userconfig.RealtimeAPIKind {
+		return
+	}
+
+	apiName := apiMetadata.apiName
+
+	a.logger.Debugw("adding new api activator", zap.String("apiName", apiName))
+
+	a.activatorsMux.Lock()
+	a.apiActivators[apiName] = newAPIActivator(apiMetadata.maxQueueLength, apiMetadata.maxConcurrency)
```

Comment:
Do you think it would be safer to check if an activator already exists in the `a.apiActivators` map before creating a new one?